bundles/auth/org.eclipse.smarthome.auth.oauth2client/src/main/java/org/eclipse/smarthome/auth/oauth2client/internal/OAuthConnector.java

```diff
@@ -0,0 +1,363 @@
+                logger.debug("Oauth request parameter {}, value {}", parameters[i], parameters[i + 1]);
+                fields.add(parameters[i], parameters[i + 1]);
+            }
+        }
+        return fields;
+    }
+
+    private AccessTokenResponse doRequest(final String grantType, HttpClient httpClient, final Request request,
+            Fields fields) throws OAuthResponseException, OAuthException, IOException {
+
+        try {
+            final FormContentProvider entity = new FormContentProvider(fields);
+            final ContentResponse response = AccessController
+                    .doPrivileged((PrivilegedExceptionAction<ContentResponse>) () -> {
+                        Request requestWithContent = request.content(entity);
+                        return requestWithContent.send();
+                    });
+
+            int statusCode = response.getStatus();
+            String content = response.getContentAsString();
+
+            if (statusCode == HttpStatus.OK_200) {
+                AccessTokenResponse jsonResponse = gson.fromJson(content, AccessTokenResponse.class);
+                jsonResponse.setCreatedOn(LocalDateTime.now()); // this is not supplied by the response
+                logger.info("grant type {} to URL {} success", grantType, request.getURI());
+                return jsonResponse;
+            } else if (statusCode == HttpStatus.BAD_REQUEST_400) {
+                OAuthResponseException errorResponse = gson.fromJson(content, OAuthResponseException.class);
+                logger.error("grant type {} to URL {} failed with error code {}, description {}", grantType,
+                        request.getURI(), errorResponse.getError(), errorResponse.getErrorDescription());
+
+                throw errorResponse;
+            } else {
+                logger.error("grant type {} to URL {} failed with HTTP response code {}", grantType, request.getURI(),
+                        statusCode);
+                throw new OAuthException("Bad http response, http code " + statusCode);
+            }
+        } catch (PrivilegedActionException pae) {
+            Exception underlyingException = pae.getException();
+            if (underlyingException instanceof InterruptedException || underlyingException instanceof TimeoutException
+                    || underlyingException instanceof ExecutionException) {
+                throw new IOException("Exception in oauth communication, grant type " + grantType, underlyingException);
+            }
+            // Dont know what exception it is, wrap it up and throw it out
+            throw new OAuthException("Exception in oauth communication, grant type " + grantType, underlyingException);
+        }
+    }
+
+    private HttpClient createHttpClient(@NonNull String tokenUrl) {
+        HttpClient httpClient = httpClientFactory.createHttpClient(HTTP_CLIENT_CONSUMER_NAME, tokenUrl);
```

Comment:
The httpClient needs to be a custom one, to take advantage of trust manager certificate pinning on a as-needed basis.  However, the Oauth requests are short lived -- there may be just 1 request for the token, and then it will not be used for a very very long time (hours/ days) until the token expires.

Therefore, it seems fit to use it once, and then shutdown after.  There may be small benefit to implement caching here but drives the complexity up much higher.

Reply:
ok, would be great if you could comment that in the code!